pkg/cvo/status_history.go

```diff
@@ -0,0 +1,235 @@
+package cvo
+
+import (
+	"math"
+	"strings"
+
+	"k8s.io/klog/v2"
+
+	configv1 "github.com/openshift/api/config/v1"
+)
+
+const (
+	// maxFinalEntryIndex is an upper bound index. History entries from 0 to maxFinalEntryIndex are given mostImportantWeight.
+	maxFinalEntryIndex = 4
+
+	// mostImportantWeight is given to history entries that should never be removed including the initial entry,
+	// the most recent entries (indices 0..maxFinalEntryIndex), and the most recently completed update.
+	mostImportantWeight = 1000.0
+
+	// interestingWeight is given to the first or last completed update in a given minor. These are interesting
+	// but not critical.
+	interestingWeight = 30.0
+
+	// partialMinorWeight is given to any partial minor updates between a completed minor version transition.
+	partialMinorWeight = 20.0
+
+	// partialZStreamWeight is given to any partial z-stream updates between a completed z-stream version transition.
+	partialZStreamWeight = -20.0
+
+	// sliceIndexWeight when applied will favor more recent updates and avoid ties.
+	sliceIndexWeight = -1.01
+)
+
+// maxHistorySize is set by the function prune and used by all the history pruning functions. Setting and using
+// this local package variable eases unit testing by allowing testing against a smaller history slice.
+var maxHistorySize int
+
+// prune prunes history when at maxSize by ranking each entry using the above defined weights and removing the entry with
+// the lowest rank. maxSize is passed in to allow the use of smaller values which eases unit test.
+func prune(history []configv1.UpdateHistory, maxSize int) []configv1.UpdateHistory {
+	if len(history) <= maxSize {
+		return history
+	}
+	maxHistorySize = maxSize
+
+	idx := getTheMostRecentCompletedEntryIndex(history)
```

Comment:
nit: `idx` for `index` is pretty generic.  Can we use `mostRecentCompletedEntryIndex` or something super-obvious here?  Less room for someone to misunderstand what it's about later, if they come in here and start shuffling code.

Reply:
agree